enclave/src/accounts_nonce_storage.rs

```diff
@@ -0,0 +1,406 @@
+                }
+            } else {
+                error!("StorageProofChecker returned None");
+                return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+            }
+        } else {
+            error!("Linkedlist is broken");
+            return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+        }
+    }
+
+    Ok(())
+}
+
+pub fn storage_map_key<K: Encode>(
+    module_prefix: &str,
+    storage_prefix: &str,
+    mapkey1: &K,
+    hasher1: &StorageHasher,
+) -> Vec<u8> {
+    let mut bytes = sp_core::twox_128(module_prefix.as_bytes()).to_vec();
+    bytes.extend(&sp_core::twox_128(storage_prefix.as_bytes())[..]);
+    bytes.extend(key_hash(mapkey1, hasher1));
+    bytes
+}
+
+/// generates the key's hash depending on the StorageHasher selected
+fn key_hash<K: Encode>(key: &K, hasher: &StorageHasher) -> Vec<u8> {
+    let encoded_key = key.encode();
+    match hasher {
+        StorageHasher::Identity => encoded_key.to_vec(),
+        StorageHasher::Blake2_128 => sp_core::blake2_128(&encoded_key).to_vec(),
+        StorageHasher::Blake2_128Concat => {
+            // copied from substrate Blake2_128Concat::hash since StorageHasher is not public
+            let x: &[u8] = encoded_key.as_slice();
+            sp_core::blake2_128(x)
+                .iter()
+                .chain(x.iter())
+                .cloned()
+                .collect::<Vec<_>>()
+        }
+        StorageHasher::Blake2_256 => sp_core::blake2_256(&encoded_key).to_vec(),
+        StorageHasher::Twox128 => sp_core::twox_128(&encoded_key).to_vec(),
+        StorageHasher::Twox256 => sp_core::twox_256(&encoded_key).to_vec(),
+        StorageHasher::Twox64Concat => sp_core::twox_64(&encoded_key).to_vec(),
+    }
+}
+
+pub fn create_in_memory_accounts_and_nonce_storage(
+    accounts: Vec<PolkadexAccount>,
+) -> Result<(), GatewayError> {
+    let storage = AccountsNonceStorage::create(accounts);
+    let storage_ptr = Arc::new(SgxMutex::<AccountsNonceStorage>::new(storage));
+    let ptr = Arc::into_raw(storage_ptr);
+    GLOBAL_ACCOUNTS_AND_NONCE_STORAGE.store(ptr as *mut (), Ordering::SeqCst);
+    Ok(())
+}
+
+pub struct AccountsNonceStorage {
+    pub accounts_storage: PolkadexAccountsStorage,
+    pub nonce_storage: PolkadexNonceStorage,
+}
+
+impl AccountsNonceStorage {
+    fn create(accounts: Vec<PolkadexAccount>) -> Self {
+        Self {
+            accounts_storage: PolkadexAccountsStorage::create(accounts),
+            nonce_storage: PolkadexNonceStorage::create(),
+        }
+    }
+
+    fn initialize_main_account(&mut self, acc: AccountId) -> Result<(), AccountRegistryError> {
+        self.accounts_storage.add_main_account(acc.clone())?;
+        self.nonce_storage.initialize_nonce(acc);
+        Ok(())
+    }
+
+    fn initialize_proxy_account(
+        &mut self,
+        acc: AccountId,
+        proxy: AccountId,
+    ) -> Result<(), AccountRegistryError> {
+        self.accounts_storage.add_proxy(acc, proxy)?;
```

Comment:
Since the proxy accounts are tied to a main account, I'm using the main account's nonce for validation, should i implement nonce for the proxy accounts independently?

Reply:
Yes, the proxy is the account that signs the trade so that attack can use it for replay attacks. We need nonce for proxy accounts and no on the main account. 

Think about it like that, any account that signs a transaction needs a nonce to avoid replay attacks

Reply:
Added nonce to the proxy accounts.